system/security/plugins/jwtSecurity/jwtEndpoint.cpp

```diff
@@ -0,0 +1,192 @@
+
+struct MemoryStruct
+{
+    char*   memory;
+    size_t  size;
+    size_t  capacity;
+};
+
+void InitMemoryStructCapacity(MemoryStruct& mem, size_t initCapacity)
+{
+    mem.memory = static_cast<char*>(malloc(initCapacity));
+    mem.size = 0;
+    mem.capacity = initCapacity;
+}
+
+void FreeMemoryStruct(MemoryStruct& mem)
+{
+    if (mem.memory)
+    {
+        free(mem.memory);
+        mem.memory = nullptr;
+    }
+    mem.size = 0;
+    mem.capacity = 0;
+}
+
+size_t CaptureIncomingCURLReply(void* contents, size_t size, size_t nmemb, void* userp)
+{
+    size_t          incomingDataSize = size * nmemb;
+    MemoryStruct*   mem = static_cast<struct MemoryStruct*>(userp);
+    size_t          MAX_BUFFER_SIZE = 4194304; // 2^22
+
+    if ((mem->size + incomingDataSize) < MAX_BUFFER_SIZE)
+    {
+        if ((mem->size + incomingDataSize) > mem->capacity)
+        {
+            size_t  newCapacity = mem->capacity * 2;
+
+            // Keep doubling capacity until it is greater than what we need
+            while ((mem->size + incomingDataSize) > newCapacity)
+                newCapacity *= 2;
+
+            mem->memory = static_cast<char*>(realloc(mem->memory, newCapacity));
+            mem->capacity = newCapacity;
+        }
+
+        memcpy(&(mem->memory[mem->size]), contents, incomingDataSize);
+        mem->size += incomingDataSize;
+    }
+    else
+    {
+        // Signals an error to libcurl
+        incomingDataSize = 0;
+
+        PROGLOG("CJwtSecurityManager::CaptureIncomingCURLReply() exceeded buffer size %ld", MAX_BUFFER_SIZE);
+    }
+
+    return incomingDataSize;
+}
+
+std::string _tokenFromEndpoint(const std::string& jwtEndPoint, bool allowSelfSignedCert, const std::string& credentialsStr)
+{
+    std::string     apiResponse;
+
+    // Call JWT endpoint
+    CURL*   curlHandle = curl_easy_init();
+    if (curlHandle)
+    {
+        CURLcode                curlResponseCode;
+        struct curl_slist*      headers = nullptr;
+        MemoryStruct            captureBuffer;
+        size_t                  INITIAL_BUFFER_SIZE = 32768; // 2^15
+        char                    curlErrBuffer[CURL_ERROR_SIZE];
+
+        curlErrBuffer[0] = '\0';
+
+        try
+        {
+            // Initialize our capture buffer to a reasonable size to avoid
+            // memory reallocation
+            InitMemoryStructCapacity(captureBuffer, INITIAL_BUFFER_SIZE);
+
+            headers = curl_slist_append(headers, "Content-Type: application/json;charset=UTF-8");
+            curl_easy_setopt(curlHandle, CURLOPT_HTTPHEADER, headers);
+
+            curl_easy_setopt(curlHandle, CURLOPT_URL, jwtEndPoint.c_str());
+            curl_easy_setopt(curlHandle, CURLOPT_POST, 1);
+            curl_easy_setopt(curlHandle, CURLOPT_POSTFIELDS, credentialsStr.c_str());
+            if (allowSelfSignedCert)
+                curl_easy_setopt(curlHandle, CURLOPT_SSL_VERIFYPEER, 0);
+            curl_easy_setopt(curlHandle, CURLOPT_FOLLOWLOCATION, 1);
```

Comment:
Do you find these settings are necessary for practical reasons?  Seems like not passing credentials somewhere else would be more secure.

Reply:
CURLOPT_UNRESTRICTED_AUTH is actually not needed; it is used only for basic auth, which we are not using.  That's what I get for copying and pasting from another project.  I've removed this one.

CURLOPT_FOLLOWLOCATION is needed to handle 3xx redirects used in some load balancer situations.

Are there other settings of concern?

Reply:
I don't see any others that concern me.